### PR TITLE
Spire controller manager upgrade

### PR DESCRIPTION
**Pull request by kfox1111**

### CODE REVIEW

**Comment by faisal-memon** [comment]
Thanks @kfox1111. Was just going to start creating issues for the logistical components of this refactor. Took a quick pass. The reorg and new crd look good. Seems like the flag is not wired up yet?

**Comment by kfox1111** [comment]
> Thanks @kfox1111. Was just going to start creating issues for the logistical components of this refactor. Took a quick pass. The reorg and new crd look good. Seems like the flag is not wired up yet?

Hit a snag that we discussed a bit on the meeting today. @faisal-memon is going to file a github issue for the CRD upgrade issue and we'll discuss the options there.

File: charts/spire-crds/templates/spire.spiffe.io_controllermanagerconfigs.yaml [hunk 1]
```diff
@@ -4,6 +4,7 @@ kind: CustomResourceDefinition
 metadata:
   annotations:
     controller-gen.kubebuilder.io/version: v0.8.0
+    {{ .Values.annotations | toYaml | nindent 4 }}
```
Comment (marcofranssen) [inline charts/spire-crds/templates/spire.spiffe.io_controllermanagerconfigs.yaml:7]:
```suggestion
    {{- .Values.annotations | toYaml | nindent 4 }}
```

File: charts/spire/README.md [hunk 2]
```diff
@@ -83,6 +83,38 @@ Now you can interact with the Spire agent socket from your own application. The
...
 
+## Install notes
+
+To do a quick non production install:
+
+kubectl create namespace spire-server
+helm install -n spire-server spire-crds charts/spire-crds
+helm install -n spire-server spire charts/spire
+
+For production installs, please see examples/production/
```
Comment (marcofranssen) [inline charts/spire/README.md:94]:
```suggestion
For production installs, please see [the production example](examples/production/).
```

File: charts/spire/README.md [hunk 3]
```diff
@@ -83,6 +83,38 @@ Now you can interact with the Spire agent socket from your own application. The
...
+
+## Upgrade notes
+
+0.14.X:
+
+If coming from a chart version before 0.14.0, you must relabel your crds to switch to using the new spire-crds chart. To migrate to the spire-crds chart
+run the following:
+
+```
```
Comment (marcofranssen) [inline charts/spire/README.md:103]:
```suggestion
```shell
```

File: charts/spire/README.md [hunk 4]
```diff
@@ -83,6 +83,38 @@ Now you can interact with the Spire agent socket from your own application. The
...
+kubectl annotate crd "clusterfederatedtrustdomains.spire.spiffe.io" "meta.helm.sh/release-namespace=spire-server"
+kubectl label crd "clusterspiffeids.spire.spiffe.io" "app.kubernetes.io/managed-by=Helm"
+kubectl annotate crd "clusterspiffeids.spire.spiffe.io" "meta.helm.sh/release-name=spire-crds"
+kubectl annotate crd "clusterspiffeids.spire.spiffe.io" "meta.helm.sh/release-namespace=spire-server"
+kubectl label crd "controllermanagerconfigs.spire.spiffe.io" "app.kubernetes.io/managed-by=Helm"
+kubectl annotate crd "controllermanagerconfigs.spire.spiffe.io" "meta.helm.sh/release-name=spire-crds"
+kubectl annotate crd "controllermanagerconfigs.spire.spiffe.io" "meta.helm.sh/release-namespace=spire-server"
+
+#Update -n with the namespace your installing the spire chart into.
+helm install -n spire-server spire-crds charts/spire-crds
```
Comment (marcofranssen) [inline charts/spire/README.md:115]:
```suggestion
# Set the namespace to the namespace you are installing the spire chart.
namespace=spire-system
kubectl label crd "clusterfederatedtrustdomains.spire.spiffe.io" "app.kubernetes.io/managed-by=Helm"
kubectl annotate crd "clusterfederatedtrustdomains.spire.spiffe.io" "meta.helm.sh/release-name=spire-crds"
kubectl annotate crd "clusterfederatedtrustdomains.spire.spiffe.io" "meta.helm.sh/release-namespace=${namespace}"
kubectl label crd "clusterspiffeids.spire.spiffe.io" "app.kubernetes.io/managed-by=Helm"
kubectl annotate crd "clusterspiffeids.spire.spiffe.io" "meta.helm.sh/release-name=spire-crds"
kubectl annotate crd "clusterspiffeids.spire.spiffe.io" "meta.helm.sh/release-namespace=spire-server"
kubectl label crd "controllermanagerconfigs.spire.spiffe.io" "app.kubernetes.io/managed-by=Helm"
kubectl annotate crd "controllermanagerconfigs.spire.spiffe.io" "meta.helm.sh/release-name=spire-crds"
kubectl annotate crd "controllermanagerconfigs.spire.spiffe.io" "meta.helm.sh/release-namespace=${namespace}"

helm install -n "${namespace}" spire-crds charts/spire-crds
```

Introduced a variable for the namespace as it is also used in the annotations, which also has to be aligned.

File: charts/spire/README.md [hunk 5]
```diff
@@ -83,6 +83,38 @@ Now you can interact with the Spire agent socket from your own application. The
 | file://./charts/spire-server | spire-server | 0.1.0 |
 | file://./charts/tornjak-frontend | tornjak-frontend | 0.1.0 |
 
+## Install notes
+
+To do a quick non production install:
+
+kubectl create namespace spire-server
+helm install -n spire-server spire-crds charts/spire-crds
+helm install -n spire-server spire charts/spire
```
Comment (marcofranssen) [inline charts/spire/README.md:92]:
```suggestion
```shell
kubectl create namespace spire-system
helm install -n spire-system spire-crds charts/spire-crds
helm install -n spire-system spire charts/spire
```
```

File: .github/workflows/helm-chart-ci.yaml [hunk 6]
```diff
@@ -170,7 +170,8 @@ jobs:
 
       - name: Run chart-testing (install)
         run: |
-          ct install --config ct.yaml \
+          helm install spire-crds charts/spire-crds
```
Comment (marcofranssen) [inline .github/workflows/helm-chart-ci.yaml:173]: Should we define the namespace here?

File: .github/workflows/helm-chart-ci.yaml [hunk 7]
```diff
@@ -238,7 +239,9 @@ jobs:
           verbosity: 1
 
       - name: Install and test example
-        run: ${{ matrix.example }}/run-tests.sh
+        run: |
+          helm install spire-crds charts/spire-crds
```
Comment (marcofranssen) [inline .github/workflows/helm-chart-ci.yaml:243]: Should we define the namespace here?

**Review by marcofranssen — CHANGES_REQUESTED** [review]
Looks pretty nice and simple.

Few suggestions on documentation.

File: charts/spire-crds/templates/spire.spiffe.io_clusterfederatedtrustdomains.yaml [hunk 8]
```diff
@@ -3,7 +3,8 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.8.0
+    controller-gen.kubebuilder.io/version: v0.11.1
+    {{ .Values.annotations | toYaml | nindent 4 }}
```
Comment (marcofranssen) [inline charts/spire-crds/templates/spire.spiffe.io_clusterfederatedtrustdomains.yaml:7]:
```suggestion
    {{- .Values.annotations | toYaml | nindent 4 }}
```

File: charts/spire-crds/templates/spire.spiffe.io_clusterspiffeids.yaml [hunk 9]
```diff
@@ -3,7 +3,8 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.8.0
+    controller-gen.kubebuilder.io/version: v0.11.1
+    {{ .Values.annotations | toYaml | nindent 4 }}
```
Comment (marcofranssen) [inline charts/spire-crds/templates/spire.spiffe.io_clusterspiffeids.yaml:7]:
```suggestion
    {{- .Values.annotations | toYaml | nindent 4 }}
```

File: charts/spire-crds/templates/spire.spiffe.io_clusterstaticentries.yaml [hunk 10]
```diff
@@ -0,0 +1,92 @@
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.11.1
+    {{ .Values.annotations | toYaml | nindent 4 }}
```
Comment (marcofranssen) [inline charts/spire-crds/templates/spire.spiffe.io_clusterstaticentries.yaml:7]:
```suggestion
    {{- .Values.annotations | toYaml | nindent 4 }}
```

File: .github/workflows/helm-chart-ci.yaml (hunk 6 above)
Comment (kfox1111) [reply .github/workflows/helm-chart-ci.yaml:173]: We could. Would have to add a kubectl create namespace here too. I figured for the tests it probably didn't matter too much?

File: charts/spire/README.md (hunk 4 above)
Comment (kfox1111) [reply charts/spire/README.md:115]: This one is a bit trickier to deal with...

I spelled it all out,so the commands would work no matter which os you use. With the variables, then it only works with linux/macos/windows but only with bash. If the user is using cmd/powershell, it breaks.

Should we have them edit the strings in place and then only need one set of instructions, or have two, and include a powershell example?

File: charts/spire/charts/spire-server/charts/spire-crds/README.md [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+| marcofranssen | <marco.franssen@gmail.com> | <https://marcofranssen.nl> |
+| kfox1111 | <Kevin.Fox@pnnl.gov> |  |
+| faisal-memon | <fymemon@yahoo.com> |  |
+| edwbuck | <edwbuck@gmail.com> |  |
+
+## Source Code
+
+* <https://github.com/spiffe/helm-charts/tree/main/charts/spire>
+
+## Parameters
```
[the remark this answers is not in the document]
Comment (kfox1111) [reply charts/spire/charts/spire-server/charts/spire-crds/README.md]: Couldn't merge due to chart location change in the repo. I think I manually merged it ok.

File: .github/workflows/helm-chart-ci.yaml (hunk 7 above)
Comment (kfox1111) [reply .github/workflows/helm-chart-ci.yaml]: Done.

File: .github/workflows/helm-chart-ci.yaml (hunk 6 above)
Comment (kfox1111) [reply .github/workflows/helm-chart-ci.yaml]: Done

File: charts/spire/README.md (hunk 4 above)
Comment (marcofranssen) [reply charts/spire/README.md:116]: OK. In that case maybe just move the comment to the top to cover the whole block.

```suggestion
# replace the spire-server namespace in below commands with the namespace you have been using to install spire-server.

kubectl label crd "clusterfederatedtrustdomains.spire.spiffe.io" "app.kubernetes.io/managed-by=Helm"
kubectl annotate crd "clusterfederatedtrustdomains.spire.spiffe.io" "meta.helm.sh/release-name=spire-crds"
kubectl annotate crd "clusterfederatedtrustdomains.spire.spiffe.io" "meta.helm.sh/release-namespace=spire-server"
kubectl label crd "clusterspiffeids.spire.spiffe.io" "app.kubernetes.io/managed-by=Helm"
kubectl annotate crd "clusterspiffeids.spire.spiffe.io" "meta.helm.sh/release-name=spire-crds"
kubectl annotate crd "clusterspiffeids.spire.spiffe.io" "meta.helm.sh/release-namespace=spire-server"
kubectl label crd "controllermanagerconfigs.spire.spiffe.io" "app.kubernetes.io/managed-by=Helm"
kubectl annotate crd "controllermanagerconfigs.spire.spiffe.io" "meta.helm.sh/release-name=spire-crds"
kubectl annotate crd "controllermanagerconfigs.spire.spiffe.io" "meta.helm.sh/release-namespace=spire-server"
helm install -n spire-server spire-crds charts/spire-crds
```

File: charts/spire/README.md [hunk 12]
```diff
@@ -83,6 +83,39 @@ Now you can interact with the Spire agent socket from your own application. The
...
+
+## Upgrade notes
+
+0.14.X:
+
+If coming from a chart version before 0.14.0, you must relabel your crds to switch to using the new spire-crds chart. To migrate to the spire-crds chart
+run the following:
+
+```shell
+# replace the spire-server namespace in below commands with the namespace you have been using to install spire-server.
```
Comment (kfox1111) [inline charts/spire/README.md:105]:
```suggestion
# Replace the spire-server namespace in the commands below with the namespace you want to install the spire-crds chart in.
```

**Review by marcofranssen — APPROVED** [review]
LGTM 🚀

**Pull request merged**